### PR TITLE
8108 zdb -l fails to read labels 2 and 3

### PR DESCRIPTION
**Pull request by yuripv**
I've introduced a regression in ```zdb -l``` while implementing #6866 - st_size is not defined for character special files, so we always had psize being 0 - fix the logic a bit to do a fstat64() after opening the device.

While here, make slice "autocompletion" a bit smarter, checking for p* as well, so if we specified ```zdb -l c0t0d0p0```, we'll use ```/dev/rdsk/c0t0d0p0``` and not ```/dev/rdsk/c0t0d0p0s0```.

Both reported by Toomas Soome.

### CODE REVIEW

File: usr/src/cmd/zdb/zdb.c [hunk 1]
```diff
@@ -2301,11 +2301,18 @@ dump_label(const char *dev)
 
 		(void) snprintf(path, sizeof (path), "%s%s", ZFS_RDISK_ROOTD,
 		    dev);
-		if ((s = strrchr(dev, 's')) == NULL || !isdigit(*(s + 1)))
+		if (((s = strrchr(dev, 's')) == NULL &&
+		    (s = strchr(dev, 'p')) == NULL) ||
+		    !isdigit(*(s + 1)))
 			(void) strlcat(path, "s0", sizeof (path));
```
Comment (tsoome) [inline usr/src/cmd/zdb/zdb.c]: I am feeling a bit odd about just using "s0" there, right now it is the case of course. with "7446 zpool create should support efi system partition" it could be "s1" as well, but 7446 has not landed. But I wonder if we should be a bit smarter there..  in any case, this is small nit IMO and not an blocker.

Comment (yuripv) [inline usr/src/cmd/zdb/zdb.c]: This is only about using a vdev name like ```c0t0d0``` - the "wholedisks" pools use ```s0```, so it's mostly used at the moment, and this doesn't stop you from providing full path, or specifying ```c0t0d0s1```.

Comment (tsoome) [inline usr/src/cmd/zdb/zdb.c]: Yep, thats the reason I do not consider it an blocker.

**Comment by ahrens** [comment]
@zettabot go

**Comment by yuripv** [comment]
Ping :)

File: usr/src/cmd/zdb/zdb.c [hunk 2]
```diff
@@ -2301,11 +2301,18 @@ dump_label(const char *dev)
...
 	}
 
-	if (stat64(path, &statbuf) != 0) {
+	if ((fd = open64(path, O_RDONLY)) < 0) {
+		(void) printf("cannot open '%s': %s\n", path, strerror(errno));
+		exit(1);
+	}
+
+	if (fstat64(fd, &statbuf) != 0) {
 		(void) printf("failed to stat '%s': %s\n", path,
```
Comment (gusev-vitaliy) [inline usr/src/cmd/zdb/zdb.c]: Please add close(fd). It is not so mandatory because exit() is called after, but it would be good to have all things cleanup-ed.

Comment (yuripv) [inline usr/src/cmd/zdb/zdb.c]: Done (and for the case below).

Comment (gusev-vitaliy) [inline usr/src/cmd/zdb/zdb.c]: Thanks.

File: usr/src/cmd/zdb/zdb.c [hunk 3]
```diff
@@ -2301,24 +2301,29 @@ dump_label(const char *dev)
...
 			(void) strlcat(path, "s0", sizeof (path));
 	}
 
-	if (stat64(path, &statbuf) != 0) {
-		(void) printf("failed to stat '%s': %s\n", path,
+	if ((fd = open64(path, O_RDONLY)) < 0) {
+		(void) fprintf(stderr, "cannot open '%s': %s\n", path,
 		    strerror(errno));
 		exit(1);
 	}
```
Comment (gusev-vitaliy) [inline usr/src/cmd/zdb/zdb.c]: It would be nice to have any proof reference in comment why fstat64() is required here and not stat64(). POSIX and man page don't describe difference in behaviour between those functions for char devices:
    st_size
                     For regular files, this is the address of the end of the
                     file. For block special or character special, this is not
                     defined. See also pipe(2).

And therefore comment about using fstat64() here is strongly recommended.

Comment (yuripv) [inline usr/src/cmd/zdb/zdb.c]: This isn't really related to zdb code or functionality - if we really want to document this better, it should be done in stat(2).

Comment (gusev-vitaliy) [inline usr/src/cmd/zdb/zdb.c]: Yes, you are right, but this comment would help to developers do not introduce regressions :)

**Comment by yuripv** [comment]
Can we get this integrated please? I've addressed all comments that are related to this specific issue.

File: usr/src/cmd/zdb/zdb.c (hunk 3 above)
Comment (gusev-vitaliy) [inline usr/src/cmd/zdb/zdb.c]: Please (!!!) add comment with reference to prove fstat behaviour.

Comment (yuripv) [inline usr/src/cmd/zdb/zdb.c]: That would serve absolutely no purpose - it's my fault for not reading the stat(2) man page correctly and is not related to zdb code at all.

File: usr/src/cmd/zdb/zdb.c (hunk 1 above)
Comment (andy-js) [inline usr/src/cmd/zdb/zdb.c]: I'm not entirely sure what this fixes.  It seems to me that it would prevent "s0" from being appended to fdisk partitions, which would just fail anyway.  Why not also check for ":" which is used by lofi then?

File: usr/src/cmd/zdb/zdb.c (hunk 3 above)
Comment (gusev-vitaliy) [inline usr/src/cmd/zdb/zdb.c]: Sorry, what information did you get from man stat(2) that describes this behaviour for character  files?

Comment (andy-js) [inline usr/src/cmd/zdb/zdb.c]: The value returned by fstat() is also wrong. It should return partition-end - partition-start (for the size), instead it just returns partition-end.  This is unlikely to cause problems here but should be documented.

Comment (yuripv) [inline usr/src/cmd/zdb/zdb.c]:
```
       st_size
                     For regular files, this is the address of the end of the
                     file. For block special or character special, this is not
                     defined. See also pipe(2).
```

File: usr/src/cmd/zdb/zdb.c (hunk 1 above)
Comment (yuripv) [inline usr/src/cmd/zdb/zdb.c]: > it would prevent "s0" from being appended to fdisk partitions

Exactly. Will it fail or not isn't related to this change.

File: usr/src/cmd/zdb/zdb.c (hunk 3 above)
Comment (gusev-vitaliy) [inline usr/src/cmd/zdb/zdb.c]: Yes, I also pointed to this. Therefore, you weren't wrong and it wasn't your fault. If just follow the man stat(2) you safely can mix stat/fstat usage w/o any problem. And comment that I ask should be as warning about behaviour that fstat() is not as stat().

Comment (yuripv) [inline usr/src/cmd/zdb/zdb.c]: I understand your concerns, but this is out of scope of this fix. Could you please add the comments you want in a separate issue?

Comment (gusev-vitaliy) [inline usr/src/cmd/zdb/zdb.c]: I believe that this is scope of this fix because you got this regression (and reviewers) exactly by the same reason: no comment about this difference.

Comment (yuripv) [inline usr/src/cmd/zdb/zdb.c]: OK, I hear you, but choose not to do that. It's up to advocates to accept this or not.

Comment (gusev-vitaliy) [inline usr/src/cmd/zdb/zdb.c]: Ok :)

File: usr/src/cmd/zdb/zdb.c (hunk 1 above)
Comment (andy-js) [inline usr/src/cmd/zdb/zdb.c]: Right.  And this change is not related to the label issue.

Comment (yuripv) [inline usr/src/cmd/zdb/zdb.c]: Indeed, but I choose to include it here as it was part of my work.

**Comment by prakashsurya** [comment]
I'll open the RTI for this today.

**Comment by yuripv** [comment]
@prakashsurya please mention that @gusev-vitaliy and @andy-js had some comments that I chose not to implement so that there are no further questions.

**Comment by andy-js** [comment]
Thanks Yuri.  Feel free to count me as a reviewer.
